File: dsa_leetcode_beginner/0190-reverse-bits/0190-reverse-bits.py
class Solution:
    def reverseBits(self, n: int) -> int:
        result = 0

        for _ in range(32):
            # Extract the last bit of n
            bit = n & 1

            # Shift result to the left to make space for the extracted bit
            result = (result << 1) | bit

            # Shift n to the right to process the next bit
            n = n >> 1

        return result
    

#Question: https://leetcode.com/problems/reverse-bits
#Blog: https://blog.unwiredlearning.com/reverse-bits

File: dsa_leetcode_beginner/0190-reverse-bits/0190-reverse-bits.js

function reverseBits(n) {
    let result = 0;

    for (let i = 0; i < 32; i++) {
        let bit = n & 1;  // Extract the last bit of n
        result = (result << 1) | bit;  // Shift result to the left and add bit
        n = n >>> 1;  // Logical right shift n to process the next bit
    }

    return result >>> 0;  // Convert to unsigned 32-bit integer
}

// Example usage:
let n = 43261596;  // Example input (binary: 00000010100101000001111010011100)
console.log(reverseBits(n));  // Output: 964176192 (binary: 00111001011110000010100101000000)
